### PR TITLE
Integrate segments with storefront API

### PR DESCRIPTION
**Pull request by bencrouse**

### CODE REVIEW

**Review by chriscressman — APPROVED** [review]

File: storefront/app/controllers/workarea/api/storefront/categories_controller.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@ class CategoriesController < Api::Storefront::ApplicationController
         before_action :cache_page
 
         def index
-          models = Catalog::Category.active.page(params[:page])
+          models = Catalog::Category.page(params[:page]).select(&:active?)
```
Comment (tubbo) [inline storefront/app/controllers/workarea/api/storefront/categories_controller.rb:8]: does this still paginate?

File: storefront/app/controllers/workarea/api/storefront/checkouts_controller.rb [hunk 2]
```diff
@@ -96,7 +96,6 @@ def touch_checkout
           if authentication?
             current_order.touch_checkout!(
               ip_address: request.remote_ip,
-              user_activity_id: current_user.id,
```
Comment (tubbo) [inline storefront/app/controllers/workarea/api/storefront/checkouts_controller.rb:99]: why did you remove this? it's in the normal CheckoutsController as `user_activity_id: current_user_activity_id,`

File: storefront/config/initializers/config.rb [hunk 3]
```diff
@@ -1,2 +1,21 @@
...
+#
+# Since this is before any other middleware (including Rails' routing), we don't
+# have a way to check the controller class or anything else application-related
+# for whether it's an API request or an ordinary request.
+#
+# The best thing I could come up with is this regex. This lambda provides a hook
+# for builds in case it doesn't work. They can provide their own logic here.
+#
+Workarea.config.is_api_visit = lambda do |request|
+  request.original_url =~ /:\/\/api\.|\/api\/./
```
Comment (tubbo) [inline storefront/config/initializers/config.rb:20]: personally I think a cleaner solution here would be to test for the presence of the `Authentication` header, and whether its value starts with `Token`. Since all storefront API requests are token-authenticated, and we don't use token authentication for anything else, I'm pretty sure we can rely on the fact that if an `Authorization: Token token=...` is present on an HTTP request, that it's an API request.

**edit:** if you feel that's too brittle, another solution is to use a [vendor-specific MIME type](https://en.wikipedia.org/wiki/Media_type#Vendor_tree) like `application/vnd.workarea+json`, [similar to the GitHub API](https://developer.github.com/v3/media/). The old `application/json` MIME type would be backwards-compatible, but we could warn users that segmentation would not be respected if you use the original header.

**Review by tubbo — CHANGES_REQUESTED** [review]

File: storefront/app/controllers/workarea/api/storefront/categories_controller.rb (hunk 1 above)
Comment (bencrouse) [reply storefront/app/controllers/workarea/api/storefront/categories_controller.rb:8]: Yes, but the actual results might be less than the per-page. Not sure how else we'd achieve this.

File: storefront/app/controllers/workarea/api/storefront/checkouts_controller.rb (hunk 2 above)
Comment (bencrouse) [reply storefront/app/controllers/workarea/api/storefront/checkouts_controller.rb:99]: https://github.com/workarea-commerce/workarea/blob/master/core/app/models/workarea/order.rb#L37

File: storefront/config/initializers/config.rb (hunk 3 above)
Comment (bencrouse) [reply storefront/config/initializers/config.rb:20]: Not all storefront requests are authenticated. I'd like to break functionality as little as possible.

**Review by bencrouse — COMMENTED** [review]
Github seems confused about my commenting, let's see if this fixes.

File: storefront/test/documentation/workarea/api/storefront/segmentation_documentation_test.rb [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+          description 'Specifying session count for segmentation'
+          route storefront_api.system_content_path('home_page')
+          explanation <<~EOS
+            Workarea supports segmenting users by number of sessions, e.g.
+            First-time vs returning visitors. To support this in the storefront
+            API, the client will be responsible for managing this since the API
+            is stateless (doesn't have cookies/session).
+
+            To get session-based segments functioning, you need to pass a
+            `segment_id` parameter in each request. Workarea will use that as
```
Comment (mttdffy) [inline storefront/test/documentation/workarea/api/storefront/segmentation_documentation_test.rb]: this seems wrong when looking at the test beneath? you're passing `sessions`?

**Review by mttdffy — CHANGES_REQUESTED** [review]

File: storefront/test/documentation/workarea/api/storefront/segmentation_documentation_test.rb (hunk 4 above)
Comment (bencrouse) [reply storefront/test/documentation/workarea/api/storefront/segmentation_documentation_test.rb]: Whoa what was I smoking.

File: storefront/app/controllers/workarea/api/storefront/categories_controller.rb (hunk 1 above)
Comment (tubbo) [reply storefront/app/controllers/workarea/api/storefront/categories_controller.rb:8]: that's true, but we got infinite scroll so it's all good ;)
